pkg/console/oidcsetup/oidcsetup.go

```diff
@@ -0,0 +1,262 @@
+		if err != nil {
+			return statusHandler.FlushAndReturn(err)
+		}
+	}
+
+	oidcClientsSchema, err := authnConfigHasOIDCFields(c.crdLister)
+	if err != nil {
+		return statusHandler.FlushAndReturn(err)
+	}
+
+	if oidcClientsSchema {
+		applyErr := c.authStatusHandler.Apply(ctx, authnConfig)
+		statusHandler.AddConditions(status.HandleProgressingOrDegraded("AuthStatusHandler", "FailedApply", applyErr))
+		if applyErr != nil {
+			return statusHandler.FlushAndReturn(applyErr)
+		}
+	}
+
+	return statusHandler.FlushAndReturn(nil)
+}
+
+func (c *oidcSetupController) syncAuthTypeOIDC(
+	ctx context.Context,
+	controllerContext factory.SyncContext,
+	statusHandler status.StatusHandler,
+	operatorConfig *operatorv1.Console,
+	authnConfig *configv1.Authentication,
+) error {
+
+	clientConfig := utilsub.GetOIDCClientConfig(authnConfig)
+	if clientConfig == nil {
+		c.authStatusHandler.WithCurrentOIDCClient("")
+		c.authStatusHandler.Unavailable("OIDCClientConfig", "no OIDC client found")
+		return nil
+	}
+
+	if len(clientConfig.ClientID) == 0 {
+		err := fmt.Errorf("no ID set on OIDC client")
+		statusHandler.AddConditions(status.HandleProgressingOrDegraded("OIDCClientConfig", "MissingID", err))
+		return statusHandler.FlushAndReturn(err)
+	}
+	c.authStatusHandler.WithCurrentOIDCClient(clientConfig.ClientID)
+
+	if len(clientConfig.ClientSecret.Name) == 0 {
+		c.authStatusHandler.Degraded("OIDCClientMissingSecret", "no client secret in the OIDC client config")
+		return nil
+	}
+
+	clientSecret, err := c.targetNSSecretsLister.Secrets(api.TargetNamespace).Get("console-oauth-config")
+	if err != nil {
+		c.authStatusHandler.Degraded("OIDCClientSecretGet", err.Error())
+		return err
+	}
+
+	if valid, msg, err := c.checkClientConfigStatus(authnConfig, clientSecret); err != nil {
+		c.authStatusHandler.Degraded("DeploymentOIDCConfig", err.Error())
+		return err
+
+	} else if !valid {
+		c.authStatusHandler.Progressing("DeploymentOIDCConfig", msg)
+		return nil
+	}
+
+	c.authStatusHandler.Available("OIDCConfigAvailable", "")
+	return nil
+}
+
+// checkClientConfigStatus checks whether the current client configuration is being currently in use,
+// by looking at the deployment status. It checks whether the deployment is available and updated,
+// and also whether the resource versions for the oauth secret and server CA trust configmap match
+// the deployment.
+func (c *oidcSetupController) checkClientConfigStatus(authnConfig *configv1.Authentication, clientSecret *corev1.Secret) (bool, string, error) {
+	depl, err := c.targetNSDeploymentsLister.Deployments(api.OpenShiftConsoleNamespace).Get(api.OpenShiftConsoleDeploymentName)
+	if err != nil {
+		return false, "", err
+	}
+
+	deplAvailableUpdated := deploymentsub.IsAvailableAndUpdated(depl)
+	if !deplAvailableUpdated {
+		return false, "deployment unavailable or outdated", nil
+	}
+
+	if clientSecret.GetResourceVersion() != depl.ObjectMeta.Annotations["console.openshift.io/oauth-secret-version"] {
```

Comment:
this looks questionable. ResourceVersion rarely matches since anything can do a read/write cycle or annotate for instance.

Reply:
Generation or hash is more suitable.  This might work well enough if that's how the annotation is currently set, but it's vulnerable and ought to be fixed in a z

Reply:
I'd keep it for now. We set the annotation this way and the secret is controlled by us.

Reply:
> I'd keep it for now. We set the annotation this way and the secret is controlled by us.

Needs bug to correct by 4.17.  annotations can be set by nearly anyone and we share them, so the RV isn't guaranteed to match.